src/snowflake/cli/api/connections.py

```diff
@@ -0,0 +1,206 @@
+# Copyright (c) 2024 Snowflake Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import asyncio
+import logging
+import re
+import warnings
+from dataclasses import asdict, dataclass, field, fields, replace
+from pathlib import Path
+from typing import Optional
+
+from snowflake.cli.api.config import get_default_connection_name
+from snowflake.cli.api.exceptions import InvalidSchemaError
+from snowflake.connector import SnowflakeConnection
+from snowflake.connector.compat import IS_WINDOWS
+
+logger = logging.getLogger(__name__)
+
+schema_pattern = re.compile(r".+\..+")
+
+
+@dataclass
+class ConnectionContext:
+    # FIXME: can reduce duplication using config.ConnectionConfig
+    connection_name: Optional[str] = None
+    account: Optional[str] = None
+    database: Optional[str] = None
+    role: Optional[str] = None
+    schema: Optional[str] = None
+    user: Optional[str] = None
+    password: Optional[str] = field(default=None, repr=False)
+    authenticator: Optional[str] = None
+    private_key_file: Optional[str] = None
+    warehouse: Optional[str] = None
+    mfa_passcode: Optional[str] = None
+    enable_diag: Optional[bool] = False
+    diag_log_path: Optional[Path] = None
+    diag_allowlist_path: Optional[Path] = None
+    temporary_connection: bool = False
+    session_token: Optional[str] = None
+    master_token: Optional[str] = None
+    token_file_path: Optional[Path] = None
+
+    VALIDATED_FIELD_NAMES = ["schema"]
+
+    def clone(self) -> ConnectionContext:
+        return replace(self)
+
+    def update(self, **updates):
+        """
+        Given a dictionary of property (key, value) mappings, update properties
+        of this context object with equivalent names to the keys.
+
+        Raises KeyError if a non-property is specified as a key.
+        """
+        field_map = {field.name: field for field in fields(self)}
+        for (key, value) in updates.items():
+            # ensure key represents a property
+            if key not in field_map:
+                raise KeyError(f"{key} is not a field of {self.__class__.__name__}")
+            setattr(self, key, value)
+
+    def __repr__(self) -> str:
+        """Minimal repr where empty (i.e. None) values have their keys omitted."""
```

Comment:
None and empty aren't the same thing
```suggestion
        """Minimal repr where None values have their keys omitted."""
```